src/adc.rs

```diff
@@ -0,0 +1,534 @@
+//! # Work in Progress
+//! API for the ADC (Analog to Digital Converter)
+//!
+//! Note that the more specific your hardware selection is
+//! (e.g. stm32f303**xc** instead of just stm32f303)
+//! the more functionality is accessible
+//! (in this case: ADC3 and ADC4 additionaly to ADC1 and ADC2).
+//!
+//! # Examples
+//! For a simple, working example check `adc.rs` in the examples folder.
+use cortex_m::asm;
+use embedded_hal::adc::{Channel, OneShot};
+
+use crate::rcc::{Clocks, AHB};
+
+#[cfg(feature = "stm32f303")]
+const MAX_ADVREGEN_STARTUP_US: u32 = 10;
+
+#[cfg(feature = "stm32f303")]
+use crate::gpio::{gpioa, gpiob, gpioc, Analog};
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+use crate::gpio::{gpiod, gpioe, gpiof};
+
+#[cfg(feature = "stm32f303")]
+use crate::stm32::{ADC1, ADC1_2, ADC2};
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+use crate::stm32::{ADC3, ADC3_4, ADC4};
+
+/// ADC configuration
+///
+/// TODO: Remove `pub` from the register block once all functionalities are implemented.
+/// Leave it here until then as it allows easy access to the registers.
+pub struct Adc<ADC> {
+    pub rb: ADC,
+    clocks: Clocks,
+    prescale: Prescale,
+    operation_mode: Option<OperationMode>,
+}
+
+/// ADC sampling time
+///
+/// Each channel can be sampled with a different sample time.
+/// For Sampletime T_n the total conversion time (in ADC clock cycles) is
+/// 12.5 + (n + .5)
+///
+/// TODO: there are boundaries on how this can be set depending on the hardware.
+/// Check them and implement a sample time setting mechanism.
+pub enum SampleTime {
+    T_1,
+    T_2,
+    T_4,
+    T_7,
+    T_19,
+    T_61,
+    T_181,
+    T_601,
+}
+
+impl SampleTime {
+    /// Get the default timer
+    pub fn default() -> Self {
+        SampleTime::T_19
+    }
+
+    /// Conversion to bits for SMP
+    fn bitcode(&self) -> u8 {
+        match self {
+            SampleTime::T_1 => 0b000,
+            SampleTime::T_2 => 0b001,
+            SampleTime::T_4 => 0b010,
+            SampleTime::T_7 => 0b011,
+            SampleTime::T_19 => 0b100,
+            SampleTime::T_61 => 0b101,
+            SampleTime::T_181 => 0b110,
+            SampleTime::T_601 => 0b111,
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq)]
+/// ADC operation mode
+///
+/// TODO: Implement other modes (DMA, Differential,…)
+pub enum OperationMode {
+    OneShot,
+}
+
+#[derive(Clone, Copy)]
+/// ADC prescale
+///
+/// TODO: Implement a setter that checks requirements for `HCLK_1` and add it.
+pub enum Prescale {
+    HCLK_2 = 2,
+    HCLK_4 = 4,
+}
+
+impl Prescale {
+    /// Get default prescaler
+    fn default() -> Self {
+        Prescale::HCLK_2
+    }
+
+    /// Conversion to bits for CKMODE in ADCx_CCR
+    fn bitcode(&self) -> u8 {
```

Comment:
Right now:
``` 
assert_eq!(4, <u8>::From(Prescale::HCLK_4));  # b.c. pub enum Prescale { […]￼    HCLK_4 = 4, ￼}
assert_eq!(3, Prescale::HCLK_4.bitcode());  # b.c. 0b11 <-> HCLK_4
```
This is why I also used the function `bitcode` in other situations. 

Any suggestions on how we can resolve this consistently?

Reply:
Actually, this originated in me using fmt::Debug and doesn't make sense anymore. Implementing into is probably the best idea.

Reply:
Whoops yeah, you are right on this one. However, on `SampleTime` shouldn't #[repr(u8)] be sufficient? And then implementing the [Default](https://doc.rust-lang.org/std/default/trait.Default.html) trait by hand? 

For me this seems more idiomatic, but I have not a lot of experience regarding rust, especially with idiomatic rust.

Reply:
I don't fully get your intentions. Let's dissect:
###  `#[repr(u8)]` 
As far as I understand it, this is only affecting the [Type Layout](https://doc.rust-lang.org/reference/type-layout.html#primitive-representations), i.e. we disallow the compiler to find a fitting size for each `SampleTime` object and fix the size to 8 bit. 
Of course in this scenario we'll never need more than 8 bit to discriminate between all the (fixed) variants, however I don't see why we should tell the compiler explicitly. This just creates another LOC without adding any functionality (that I can think of right now).

### implementing the Default by hand
[This is already implemented.](https://github.com/stm32-rs/stm32f3xx-hal/pull/47/files#diff-a4bc73d1077eb9a314216894b4643500R71)
However I can't justify why I set the default to 19. I think I saw it some other place and just implemented it here for good measure.

## Assuming your intentions
I assume you'd like to see something like:
```
#[repr(u8)]
pub enum SampleTime {
    T_1 = 0b000,
    T_2 = 0b001,
    // ...
    T_601 = 0b111,
``` 
Instead of implementing a `bitcode` function that returns these values.
The reason I decided for the latter was that 
a) we're already using the `bitcode` function in other scenarios here so it's more consistent with the other code
b) I wanted to avoid confusion regarding the (potentially future) calculation the actual sample time.

Reply:
Well #[repr(u8)] was the wrong blind guess then. 

I just wanted to say, that you use Prescale and SampleTime for two different value representations. Once to specify the values, which goes into the register, and then when using it to calculate timings. 

I just wanted to say, that the `From` trait would be the perfect candidate to convert between those two representations.

Reply:
I still don't get it. 
Let's collect what I think we agree on, so you can tell me what I misunderstood.
1. It makes sense to use one enum and one enum only for handling the prescaler. The same holds for the sample time.
2.  The two different value representations for the prescaler are necessary.
3. It's more intuitve to have the actual value of the prescaler as a plain integer than the bitcode. (Hence the plain definition in the enum definition and the `bitcode` function).
4. Allowing an enum to be casted to two different integer types while returing different values is a horrible idea and we should never ever do this.
5. It's more ideomatic (and less verbose) to write
```
enum A {
  first = 1,
  second = 2,
}
let new = A::first as u32;
```
than
```
enum A {
  first,
  second,
}
impl From<A> for u32 {
    fn from(a: A) -> u32 {
        match a {
          A::first => 1,
          A::second => 2,
        }
    }
}
let new = <u32>::from(A::first);
```
6. It's good practice to keep consistency. In this case: Also have a `bitcode` funciton for the sample time.

Reply:
Thank you for the detailed answer. You've convinced me. I think my method does not really contribute to the readability of the code. So let's keep it the way with `bitcode`. 

I hope the discussion was not too exhausting. I have a tendency to make things overcomplicated just for being more "idomatic" (as of the perspective of my experience level).

Thank you for this great work :)

Reply:
To be clear here: My answer was never about convincing anyone. I really appreciate the exchange of view points and you taking the time to check the code. No one here can have 5+ years of experience in Rust so let's try to learn and grow together.
On the other hand I don't think it makes sense to drag things out in a PR.

I also appreciated (and bookmarked) [your digression about debugging](https://github.com/stm32-rs/stm32f3xx-hal/pull/35), but feel that a PR is a bad place to put it (it might also be me having little experience with github)

Maybe @dfrankland agrees to activate the wiki so we can have discussions and agreements about idiomatic stm32f3xxx-code, debugging techniques and the like there.

Reply:
This would be an interesting topic to discuss in general across all stm32xxxxx-hal implementations. (so many x's 😅)

I'm in favor of establishing some kind of alignment across these implementations. 

But as of now, I don't know where to start, or if this alignment is already happening. I'll have to ask in the next embedded meeting. I'm sure they already thought about this. 

At least some kind of guideline for a hal implementation would be great.

Reply:
I totally agree. Having a guideline makes decisions so much easier :D

Are those "embedded meetings" local or virtual? I'd really like to attend (maybe after the toilet paper crisis is over).

Reply:
The [rust-embedded wg](https://github.com/rust-embedded/wg) does have a [matrix channel](#rust-embedded:matrix.org).  They do have a meeting every Tuesday 8pm CET over the matrix channel.

As there are many people, which are also involved into the stm32-rs project, this seems to be the place, to at least ask and contact members of the stm32-rs project.